ethers-core/src/types/transaction/response.rs

```diff
@@ -141,22 +141,78 @@ impl Transaction {
-
-        #[cfg(feature = "celo")]
-        self.inject_celo_metadata(&mut rlp);
-
-        rlp_opt(&mut rlp, &self.to);
-        rlp.append(&self.value);
-        rlp.append(&self.input.as_ref());
-        rlp.append(&self.v);
+        rlp.begin_unbounded_list();
+
+        match self.transaction_type {
+            // EIP-2930 (0x01)
+            Some(x) if x == U64::from(1) => {
+                rlp_opt(&mut rlp, &self.chain_id);
+                rlp.append(&self.nonce);
+                rlp_opt(&mut rlp, &self.gas_price);
+                rlp.append(&self.gas);
+
+                #[cfg(feature = "celo")]
+                self.inject_celo_metadata(&mut rlp);
+
+                rlp_opt(&mut rlp, &self.to);
+                rlp.append(&self.value);
+                rlp.append(&self.input.as_ref());
+                rlp_opt(&mut rlp, &self.access_list);
+                if let Some(chain_id) = self.chain_id {
+                    rlp.append(&normalize_v(self.v.as_u64(), U64::from(chain_id.as_u64())));
+                }
+            }
+            // EIP-1559 (0x02)
+            Some(x) if x == U64::from(2) => {
+                rlp_opt(&mut rlp, &self.chain_id);
+                rlp.append(&self.nonce);
+                rlp_opt(&mut rlp, &self.max_priority_fee_per_gas);
+                rlp_opt(&mut rlp, &self.max_fee_per_gas);
+                rlp.append(&self.gas);
+                rlp_opt(&mut rlp, &self.to);
+                rlp.append(&self.value);
+                rlp.append(&self.input.as_ref());
+                rlp_opt(&mut rlp, &self.access_list);
+                if let Some(chain_id) = self.chain_id {
+                    rlp.append(&normalize_v(self.v.as_u64(), U64::from(chain_id.as_u64())));
+                }
+            }
+            // Legacy (0x00)
+            _ => {
+                rlp.append(&self.nonce);
+                rlp_opt(&mut rlp, &self.gas_price);
+                rlp.append(&self.gas);
+
+                #[cfg(feature = "celo")]
+                self.inject_celo_metadata(&mut rlp);
+
+                rlp_opt(&mut rlp, &self.to);
+                rlp.append(&self.value);
+                rlp.append(&self.input.as_ref());
+                rlp.append(&self.v);
+            }
+        }
+
         rlp.append(&self.r);
         rlp.append(&self.s);
 
-        rlp.out().freeze().into()
+        rlp.finalize_unbounded_list();
+
+        let rlp_bytes: Bytes = rlp.out().freeze().into();
+        let mut encoded = vec![];
+        match self.transaction_type {
+            Some(x) if x == U64::from(1) => {
+                encoded.extend_from_slice(&[0x1]);
+                encoded.extend_from_slice(rlp_bytes.as_ref());
+                encoded.into()
+            }
+            Some(x) if x == U64::from(2) => {
+                encoded.extend_from_slice(&[0x2]);
+                encoded.extend_from_slice(rlp_bytes.as_ref());
+                encoded.into()
+            }
+            _ => rlp_bytes,
```

Comment:
WDYT about matching `Some(x) if x == U64::from(0)` and encoding with the 0x0 prefix, and default to `rlp_bytes` only when `self.transaction_type` is not present OR if the `legacy` feature is turned on?

Reply:
AFAIK `0x0` is reserved for new tx types, not legacy: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2718.md#backwards-compatibility

Reply:
NVM. You are right, forgot that we skip it for legacy txs: https://github.com/gakonst/ethers-rs/blob/404e9c24163c35ff4055bfa3ce3f7107d163e7c7/ethers-core/src/types/transaction/eip2718.rs#L153-L155